fdbserver/BlobManager.actor.cpp

```diff
@@ -1314,6 +1320,585 @@ ACTOR Future<Void> blobWorkerRecruiter(
+	//       any files for that request.
+
+	if (BM_DEBUG) {
+		printf("Partially deleting granule %s: deleting file keys\n", granuleId.toString().c_str());
+	}
+
+	loop {
+		try {
+			for (auto& key : deletedFileKeys) {
+				tr.clear(key);
+			}
+			wait(tr.commit());
+			break;
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+
+	if (BM_DEBUG) {
+		printf("Partially deleting granule %s: success\n", granuleId.toString().c_str());
+	}
+	return Void();
+}
+
+/*
+ * This method is used to prune the range [startKey, endKey) at (and including) pruneVersion.
+ * To do this, we do a BFS traversal starting at the active granules. Then we classify granules
+ * in the history as nodes that can be fully deleted (i.e. their files and history can be deleted)
+ * and nodes that can be partially deleted (i.e. some of their files can be deleted).
+ * Once all this is done, we finally clear the pruneIntent key, if possible, to indicate we are done
+ * processing this prune intent.
+ */
+ACTOR Future<Void> pruneRange(BlobManagerData* self, KeyRef startKey, KeyRef endKey, Version pruneVersion, bool force) {
+	if (BM_DEBUG) {
+		printf("pruneRange starting for range [%s-%s) @ pruneVersion=%lld, force=%s\n",
+		       startKey.printable().c_str(),
+		       endKey.printable().c_str(),
+		       pruneVersion);
+	}
+
+	// queue of <range, startVersion, endVersion> for BFS traversal of history
+	// TODO: consider using GranuleHistoryEntry, but that also makes it a little messy
+	state std::queue<std::tuple<KeyRange, Version, Version>> historyEntryQueue;
+
+	// stacks of <granuleId, historyKey> and <granuleId> to track which granules to delete
+	state std::vector<std::tuple<UID, KeyRef>> toFullyDelete;
+	state std::vector<UID> toPartiallyDelete;
+
+	// set of granuleIds to track which granules we have already visited in traversal
+	state std::unordered_set<UID> visited; // track which granules we have already visited in traversal
+
+	state KeyRange range(KeyRangeRef(startKey, endKey)); // range for [startKey, endKey)
+
+	// find all active granules (that comprise the range) and add to the queue
+	state KeyRangeMap<UID>::Ranges activeRanges = self->workerAssignments.intersectingRanges(range);
+
+	state Transaction tr(self->db);
+	tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+	tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+
+	state KeyRangeMap<UID>::iterator activeRange;
+	for (activeRange = activeRanges.begin(); activeRange != activeRanges.end(); ++activeRange) {
+		if (BM_DEBUG) {
+			printf("Checking if active range [%s-%s), owned by BW %s, should be pruned\n",
+			       activeRange.begin().printable().c_str(),
+			       activeRange.end().printable().c_str(),
+			       activeRange.value().toString().c_str());
+		}
+
+		// assumption: prune boundaries must respect granule boundaries
+		if (activeRange.begin() < startKey || activeRange.end() > endKey) {
```

Comment:
To confirm: this check might need to even be stronger in terms of respecting some predefined "boundaries"